### PR TITLE
Adding k8s architecture guide

### PR DESCRIPTION
**Pull request by alexmolev**

### CODE REVIEW

File: content/bestpractices/kubernetes_architecture_guide.md [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+category: Manager Architecture
+draft: false
+abstract: Kubernetes Architecture Explained
+weight: 500
+aliases: /manager_architecture/create-deployment-flow/
+---
+
+# Kubernetes Architecture Explained
+
+## Cloudify Kubernetes support consists of three main components.
```
Comment (ofercloudify) [inline content/bestpractices/kubernetes_architecture_guide.md:13]:
```suggestion
The {{< param product_name >}} Kubernetes support consists of three main components:
```

File: content/bestpractices/kubernetes_architecture_guide.md [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+* [Kubernetes Service Plugin](https://docs.cloudify.co/latest/working_with/official_plugins/orchestration/kubernetes/) - The Kubernetes plugin exposes the Kubernetes API into the Cloudify TOSCA based DSL. This will allow users to deploy a Kubernetes service by referencing directly your service template on your git repo or Helm chart. Cloudify [CI/CD integration](https://docs.cloudify.co/latest/working_with/integration/) also allows users to run Kubernetes services directly without having to create a Cloudify wrapper blueprint. In this case, the Ci/CD plugin takes care of dynamically binding the Cloudify blueprint behind the scene.
```
Comment (ofercloudify) [inline content/bestpractices/kubernetes_architecture_guide.md:15]:
```suggestion
* [Kubernetes Service Plugin]({{< relref "/working_with/official_plugins/orchestration/kubernetes/" >}}) - The Kubernetes plugin exposes the Kubernetes API into the {{< param product_name >}} TOSCA based DSL. This allows users to deploy a Kubernetes service by referencing a service template or a Helm chart in the git repo. {{< param product_name >}}  [CI/CD integration]({{< relref "/working_with/integration/" >}}) also allows users to run Kubernetes services directly without having to create a {{< param product_name >}} wrapper blueprint. In this case, the Ci/CD plugin takes care of dynamically binding the {{< param product_name >}} blueprint behind the scene.
```

File: content/bestpractices/kubernetes_architecture_guide.md [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+* [Kubernetes Cluster Blueprint](https://github.com/cloudify-community/blueprint-examples/tree/master/kubernetes) - The Kubernetes blueprint provides a pre-canned template for provisioning a Kubernetes cluster. In the case of a managed Kubernetes such as GKE or EKS, the blueprint is responsible to set up the user environment. You can view such options [here](https://github.com/cloudify-community/blueprint-examples/tree/master/kubernetes/).
```
Comment (ofercloudify) [inline content/bestpractices/kubernetes_architecture_guide.md:16]:
```suggestion
* [Kubernetes Cluster Blueprint](https://github.com/cloudify-community/blueprint-examples/tree/master/kubernetes) - The Kubernetes blueprint provides a pre-canned template for provisioning a Kubernetes cluster. In the case of a managed Kubernetes platform such as GKE or EKS, the blueprint is responsible to set up the user environment. You can view such options [here](https://github.com/cloudify-community/blueprint-examples/tree/master/kubernetes/).
```

File: content/bestpractices/kubernetes_architecture_guide.md [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+* [Kubernetes Workflow](https://docs.cloudify.co/latest/working_with/official_plugins/orchestration/kubernetes/#workflows) - The workflow engine is responsible for handling install and uninstalls service workflow as well as day 2 operations such as updating of an existing service.
```
Comment (ofercloudify) [inline content/bestpractices/kubernetes_architecture_guide.md:17]:
```suggestion
* [Kubernetes Workflow]({{< relref "/working_with/official_plugins/orchestration/kubernetes/#workflows" >}}) - The workflow engine is responsible for handling the install and uninstall service workflows as well as day 2 operations such as updating of an existing service.
```

File: content/bestpractices/kubernetes_architecture_guide.md [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+
+![kubernetes_architecture]( /images/kubernetes_architecture/kubernetes_architecture.png )
+
+## Key features
+
+- Out of the box support for all major Kubernetes Platforms on private and public clouds (OpenShift, KubeSpray, GKE, EKS, and AKS).
```
Comment (ofercloudify) [inline content/bestpractices/kubernetes_architecture_guide.md:23]:
```suggestion
- Out of the box support for all major Kubernetes platforms on private and public clouds (OpenShift, KubeSpray, GKE, EKS, and AKS).
```

File: content/bestpractices/kubernetes_architecture_guide.md [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+- Support multiple authentication methods Token- and “Kube config”- based authentication.
+- Resource state management - Block the deployment to a successful completion.
```
Comment (ofercloudify) [inline content/bestpractices/kubernetes_architecture_guide.md:25]: @natishalom I don't understand this sentence

File: content/bestpractices/kubernetes_architecture_guide.md [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+- Reload of a specific Kubernetes service - Continuous updates through reload of a particular service with a new set of input parameters.
+
+### Native Kubernetes Template support
+
+* Service Template by URL or Inline -Load a service template from an external file, git-repo, or define it inline within the blueprint itself.
+* Parameters & secret management - Pass inputs to Kubernetes modules through Cloudify secrets and render them on the fly through Jinja.
```
Comment (ofercloudify) [inline content/bestpractices/kubernetes_architecture_guide.md:31]:
```suggestion
* Parameters & secret management - Pass inputs to Kubernetes modules through {{< param product_name >}} secrets and render them on the fly through Jinja.
```

File: content/bestpractices/kubernetes_architecture_guide.md [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+* Dynamic API binding for all Kubernetes resources
+  * Built-in resource-type mapping: Deployment, Service, POD etc..
```
Comment (ofercloudify) [inline content/bestpractices/kubernetes_architecture_guide.md:33]:
```suggestion
  * Built-in resource-type mapping: Deployment, Service, POD, etc.
```

File: content/bestpractices/kubernetes_architecture_guide.md [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+  * Custom Resource mapping through generic resource node type.
+  
+### Helm 3 Template Support
+[Helm 3 plugin](https://docs.cloudify.co/latest/working_with/official_plugins/orchestration/helm/) allows you to create a Helm client for each deployment; add repositories, create releases, and easily deploy an application on any existing Kubernetes cluster.
+
+### Multi Kubernetes Cluster Support
+The use cases for multi Kubernetes cluster can be driven by different needs 
+
+* Separate clusters between application and teams.
```
Comment (ofercloudify) [inline content/bestpractices/kubernetes_architecture_guide.md:42]:
```suggestion
* Separate clusters between applications and teams.
```

File: content/bestpractices/kubernetes_architecture_guide.md [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+* Separating between development and production.
+* Multi-Cloud - supporting Kubernetes clusters on multiple clouds.
+* Avoiding lock-in - allowing portability between providers.
+* Edge / IoT - managing deployment across many distributed Kubernetes clusters on edge devices.
+
+Cloudify was designed to manage highly distributed deployments and therefore fit well as the Multi-Cluster Kubernetes service broker.
```
Comment (ofercloudify) [inline content/bestpractices/kubernetes_architecture_guide.md:48]:
```suggestion
{{< param product_name >}} was designed to manage highly distributed deployments and therefore fit well as the Multi-Cluster Kubernetes service broker.
```

File: content/bestpractices/kubernetes_architecture_guide.md [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+
+The Cloudify blueprint allows users to define multiple services across different clusters as part of the same deployment template. Users can also use the Cloudify [Service Component](https://docs.cloudify.co/latest/working_with/service_composition/) DSL to bind a particular service to a Kubernetes cluster dynamically at runtime.
```
Comment (ofercloudify) [inline content/bestpractices/kubernetes_architecture_guide.md:50]:
```suggestion
The {{< param product_name >}} blueprint allows users to define multiple services across different clusters as part of the same deployment template. Users can also use the {{< param product_name >}} [Service Component]({{< relref "/working_with/service_composition/" >}}) DSL to bind a particular service to a Kubernetes cluster dynamically at runtime.
```

File: content/bestpractices/kubernetes_architecture_guide.md [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+
+### Interoperability between Kubernetes providers as well as non-Kubernetes services.
```
Comment (ofercloudify) [inline content/bestpractices/kubernetes_architecture_guide.md:52]:
```suggestion
### Interoperability between Kubernetes providers and non-Kubernetes services.
```

File: content/bestpractices/kubernetes_architecture_guide.md [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+
+Cloudify provides a rich set of out of the box plugins that support VM, Serverless. legacy apps or any other REST service as part of the same deployment. This will allow users to create an automation blueprint that includes Kubernetes and non-Kubernetes services and use the native dependency, relationship, workflow features as with any other node types. This opens up the options to enable interoperability between Kubernetes and non-Kubernetes services in the following way:
```
Comment (ofercloudify) [inline content/bestpractices/kubernetes_architecture_guide.md:54]:
```suggestion
{{< param product_name >}} provides a rich set of out of the box plugins that support VM, Serverless, legacy apps, or any other REST service as part of the same deployment. This allows users to create an automation blueprint that includes Kubernetes and non-Kubernetes services and use the native dependency, relationship, workflow features as with any other node types. This opens up the options to enable interoperability between Kubernetes and non-Kubernetes services in the following way:
```

File: content/bestpractices/kubernetes_architecture_guide.md [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+* Interoperability between Kubernetes cluster providers - Cloudify can manage deployment across multiple Kubernetes clusters using the same deployment.
```
Comment (ofercloudify) [inline content/bestpractices/kubernetes_architecture_guide.md:55]:
```suggestion
* Interoperability between Kubernetes cluster providers - {{< param product_name >}} can manage deployment across multiple Kubernetes clusters using the same deployment.
```

File: content/bestpractices/kubernetes_architecture_guide.md [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+* Interoperability between Orchestration Platforms - Cloudify allows users to define the relationship and pass context information between Different Kubernetes clusters as well as non-Kubernetes orchestrations such as Ansible, Terraform, Cloud Formation, Azure ARM, TOSCA, templates.
```
Comment (ofercloudify) [inline content/bestpractices/kubernetes_architecture_guide.md:56]:
```suggestion
* Interoperability between Orchestration Platforms - {{< param product_name >}} allows users to define the relationship and pass context information between Different Kubernetes clusters as well as non-Kubernetes orchestrations such as Ansible, Terraform, Cloud Formation, Azure ARM, TOSCA, templates.
```

File: content/bestpractices/kubernetes_architecture_guide.md [hunk 16]
```diff
@@ -0,0 +1,101 @@
...
+* End to End Service Automation  - Enable automation of Kubernetes, SaaS-based services, Function, Legacy services under a common automation scheme.
+
+### Managed as code with built-in integration with CI/CD
+Cloudify provides built-in integration with a list of [Ci/CD tools](https://docs.cloudify.co/latest/working_with/integration/) such as Jenkins, Git Actions, Circle CI, etc.
```
Comment (ofercloudify) [inline content/bestpractices/kubernetes_architecture_guide.md:60]:
```suggestion
{{< param product_name >}} provides built-in integration with a list of [Ci/CD tools]({{< relref "/working_with/integration/" >}}) such as Jenkins, Git Actions, Circle CI, etc.
```

File: content/bestpractices/kubernetes_architecture_guide.md [hunk 17]
```diff
@@ -0,0 +1,101 @@
...
+
+This integration includes also a generic template that will allow users to run any Kubernetes services without having to write any Cloudify blueprint as a wrapper blueprint. The wrapper Cloudify blueprint will be attached to the service behind the scene and in this case, will simplify significantly the learning curve needed to run those services through Cloudify and the Ci/CD tool of choice.
```
Comment (ofercloudify) [inline content/bestpractices/kubernetes_architecture_guide.md:62]:
```suggestion
This integration includes a generic template that allows users to run any Kubernetes service without having to write a {{< param product_name >}} blueprint as a wrapper. The {{< param product_name >}} wrapper blueprint will be attached to the service behind the scene, and in this case, will simplify significantly the learning curve needed to run those services through {{< param product_name >}} and the Ci/CD tool of choice.
```

File: content/bestpractices/kubernetes_architecture_guide.md [hunk 18]
```diff
@@ -0,0 +1,101 @@
...
+
+### Management UI
+The Cloudify management UI is focused on service management and less on monitoring the Kubernetes infrastructure. As such it provides a catalog service that provides a simple interface to deploy Helm or native Kubernetes services. Topology view which tracks the service dependency and state of a deployment. Workflow which monitors the execution steps on a particular service.
```
Comment (ofercloudify) [inline content/bestpractices/kubernetes_architecture_guide.md:65]:
```suggestion
The {{< param product_name >}} management UI is focused on service management and less on monitoring the Kubernetes infrastructure. As such it provides a catalog service that provides a simple interface to deploy Helm or native Kubernetes services. Topology view which tracks the service dependency and state of a deployment. Workflow which monitors the execution steps on a particular service. 
```

File: content/bestpractices/kubernetes_architecture_guide.md [hunk 19]
```diff
@@ -0,0 +1,101 @@
...
+
+### Managed as code with built-in integration with CI/CD
+Cloudify provides built-in integration with a list of [Ci/CD tools](https://docs.cloudify.co/latest/working_with/integration/) such as Jenkins, Git Actions, Circle CI, etc.
+
+This integration includes also a generic template that will allow users to run any Kubernetes services without having to write any Cloudify blueprint as a wrapper blueprint. The wrapper Cloudify blueprint will be attached to the service behind the scene and in this case, will simplify significantly the learning curve needed to run those services through Cloudify and the Ci/CD tool of choice.
+
+### Management UI
+The Cloudify management UI is focused on service management and less on monitoring the Kubernetes infrastructure. As such it provides a catalog service that provides a simple interface to deploy Helm or native Kubernetes services. Topology view which tracks the service dependency and state of a deployment. Workflow which monitors the execution steps on a particular service. 
+
+#### Ui Component
```
Comment (ofercloudify) [inline content/bestpractices/kubernetes_architecture_guide.md:67]:
```suggestion
#### UI Component
```

File: content/bestpractices/kubernetes_architecture_guide.md [hunk 20]
```diff
@@ -0,0 +1,101 @@
...
+
+* Service Catalogue
```
Comment (ofercloudify) [inline content/bestpractices/kubernetes_architecture_guide.md:69]:
```suggestion
* Service Catalog
```

File: content/bestpractices/kubernetes_architecture_guide.md [hunk 21]
```diff
@@ -0,0 +1,101 @@
...
+
+![ServiceCatalogue]( /images/kubernetes_architecture/service_catalogue.png )
```
Comment (ofercloudify) [inline content/bestpractices/kubernetes_architecture_guide.md:71]:
```suggestion
![ServiceCatalog]( /images/kubernetes_architecture/service_catalogue.png )
```

File: content/bestpractices/kubernetes_architecture_guide.md [hunk 22]
```diff
@@ -0,0 +1,101 @@
...
+
+* Topology view - View Kubernetes task execution and their associated infrastructure resources on a single view
```
Comment (ofercloudify) [inline content/bestpractices/kubernetes_architecture_guide.md:73]:
```suggestion
* Topology view - View Kubernetes task execution and the associated infrastructure resources in a single view
```

File: content/bestpractices/kubernetes_architecture_guide.md [hunk 23]
```diff
@@ -0,0 +1,101 @@
...
+
+![TopologyView]( /images/kubernetes_architecture/topology_view.png )
+
+* Workflow and log monitoring - View the execution graph of ansible playbooks as part of the entire end to end execution.
+
+![workflow]( /images/kubernetes_architecture/workflow.png ) 
+
+## Getting Started with Cloudify Kubernetes
```
Comment (ofercloudify) [inline content/bestpractices/kubernetes_architecture_guide.md:81]:
```suggestion
## Getting Started with {{< param product_name >}} Kubernetes
```

File: content/bestpractices/kubernetes_architecture_guide.md [hunk 24]
```diff
@@ -0,0 +1,101 @@
...
+
+### Hello world example
+The first example is a very simple hello world application. The application consists of a Kubernetes Deployment type resource and a Kubernetes Service type resource. The resources are packaged in [a single YAML file](https://github.com/cloudify-community/blueprint-examples/blob/master/kubernetes/plugin-examples/hello-world/resources.yaml). This is a traditional Kubernetes resource template with one unique difference, which is that it has been parameterized using a Jinja template. This file is delivered to the Cloudify manager with a [Cloudify blueprint](https://github.com/cloudify-community/blueprint-examples/blob/master/kubernetes/plugin-examples/hello-world/blueprint.yaml). In the blueprint, a single node template refers to this file. The “container” and “container_port” parameters may be changed to replace the application and port parameters in the Kubernetes resources.
```
Comment (ofercloudify) [inline content/bestpractices/kubernetes_architecture_guide.md:84]:
```suggestion
The first example is a very simple hello world application. The application consists of a Kubernetes Deployment type resource and a Kubernetes Service type resource. The resources are packaged in [a single YAML file](https://github.com/cloudify-community/blueprint-examples/blob/master/kubernetes/plugin-examples/hello-world/resources.yaml). This is a traditional Kubernetes resource template with one unique difference, which is that it has been parameterized using a Jinja template. This file is delivered to the {{< param cfy_manager_name >}} with a [blueprint](https://github.com/cloudify-community/blueprint-examples/blob/master/kubernetes/plugin-examples/hello-world/blueprint.yaml). In the blueprint, a single node template refers to this file. The “container” and “container_port” parameters may be changed to replace the application and port parameters in the Kubernetes resources.
```

File: content/bestpractices/kubernetes_architecture_guide.md [hunk 25]
```diff
@@ -0,0 +1,101 @@
...
+
+### Multi-Cluster Example
+
+In the following example, we used a combination of  Jenkins and Cloudify [Jenkins Plugin](https://docs.cloudify.co/latest/working_with/integration/jenkins-plugin/) as a Multi Kubernetes Cluster broker between OpenShift and GKE. In this example, OpenShift act as our development environment and GKE as the production environment. The pipeline pushes the same application across the two clusters. Cloudify is used to abstract the authentication and different setup needed for each cluster environment. We refer to this pattern as EaaS (Environment as a Service).
```
Comment (ofercloudify) [inline content/bestpractices/kubernetes_architecture_guide.md:88]:
```suggestion
In the following example, we used a combination of  Jenkins and {{< param product_name >}} [Jenkins Plugin]({{< relref "/working_with/integration/jenkins-plugin/" >}}) as a Multi Kubernetes Cluster broker between OpenShift and GKE. In this example, OpenShift act as our development environment and GKE as the production environment. The pipeline pushes the same application across the two clusters. {{< param product_name >}} is used to abstract the authentication and different setup needed for each cluster environment. We refer to this pattern as EaaS (Environment as a Service).
```

File: content/bestpractices/kubernetes_architecture_guide.md [hunk 26]
```diff
@@ -0,0 +1,101 @@
...
+
+![multi_cluster_example]( /images/kubernetes_architecture/multi_cluster_example.png )
+
+Click to watch a [demo](https://vimeo.com/483013205)
+
+### Continuous Deployment Example
+Cloudify provides tight integration with popular CI/CD tools such as GitHub Actions, CircleCI, and Jenkins. This integration provides pipeline developers with the ability to involve Cloudify in their pipelines using facilities native to the CI/CD platform.
```
Comment (ofercloudify) [inline content/bestpractices/kubernetes_architecture_guide.md:95]:
```suggestion
{{< param product_name >}} provides tight integration with popular CI/CD tools such as GitHub Actions, CircleCI, and Jenkins. This integration provides pipeline developers with the ability to involve {{< param product_name >}} in their pipelines using facilities native to the CI/CD platform.
```

File: content/bestpractices/kubernetes_architecture_guide.md [hunk 27]
```diff
@@ -0,0 +1,101 @@
...
+
+For GitHub, Cloudify provides a set of GitHub Actions, one of which is the [“Create Kubernetes Environment” action](https://github.com/marketplace/actions/create-kubernetes-environment).
```
Comment (ofercloudify) [inline content/bestpractices/kubernetes_architecture_guide.md:97]:
```suggestion
For GitHub, {{< param product_name >}} provides a set of GitHub Actions, one of which is the [“Create Kubernetes Environment” action](https://github.com/marketplace/actions/create-kubernetes-environment).
```

File: content/bestpractices/kubernetes_architecture_guide.md [hunk 28]
```diff
@@ -0,0 +1,101 @@
...
+
+An example of how Cloudify’s Kubernetes GitHub action is used can be found in our [GitHub Actions example repository](https://github.com/cloudify-community/github-actions-example/).
```
Comment (ofercloudify) [inline content/bestpractices/kubernetes_architecture_guide.md:99]:
```suggestion
An example of how {{< param product_name >}}’s Kubernetes GitHub action is used can be found in our [GitHub Actions example repository](https://github.com/cloudify-community/github-actions-example/).
```

File: content/bestpractices/kubernetes_architecture_guide.md [hunk 29]
```diff
@@ -0,0 +1,101 @@
...
+### Native Kubernetes Template support
+
+* Service Template by URL or Inline -Load a service template from an external file, git-repo, or define it inline within the blueprint itself.
+* Parameters & secret management - Pass inputs to Kubernetes modules through Cloudify secrets and render them on the fly through Jinja.
+* Dynamic API binding for all Kubernetes resources
+  * Built-in resource-type mapping: Deployment, Service, POD etc..
+  * Custom Resource mapping through generic resource node type.
+  
+### Helm 3 Template Support
+[Helm 3 plugin](https://docs.cloudify.co/latest/working_with/official_plugins/orchestration/helm/) allows you to create a Helm client for each deployment; add repositories, create releases, and easily deploy an application on any existing Kubernetes cluster.
```
Comment (ofercloudify) [inline content/bestpractices/kubernetes_architecture_guide.md:37]:
```suggestion
[Helm 3 plugin]({{< relref "/working_with/official_plugins/orchestration/helm/" >}}) allows you to create a Helm client for each deployment; add repositories, create releases, and easily deploy an application on any existing Kubernetes cluster.
```

**Review by ofercloudify — CHANGES_REQUESTED** [review]

File: content/bestpractices/kubernetes_architecture_guide.md (hunk 6 above)
Comment (natishalom) [reply content/bestpractices/kubernetes_architecture_guide.md:25]: It referes to the way Kuberenets works and how our plugin augment this.
The default behavior in Kubernetes is the deployment request is async means that the call returns before the resource have completed the deployment process.

We added behavior to block the deployment request till the resource is deployed successfully

**Pull request merged**